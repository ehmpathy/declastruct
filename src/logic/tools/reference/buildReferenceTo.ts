import { DomainObject } from 'domain-objects';

import {
  DeclaredResourceReference,
  DeclaredResourceReferenceKeyType,
} from '../../../domain/DeclaredResourceReference';
import { buildPrimaryKeyTo } from './buildPrimaryKeyTo';
import { buildUniqueKeyTo } from './buildUniqueKeyTo';
import { defineReferenceClassOf } from './defineReferenceClassOf';
import { defineReferenceKeyConstituentsOf } from './defineReferenceKeyConstituentsOf';

export class CanNotReferenceDeclaredResourceClassError extends Error {
  constructor({
    class: ofClass,
    reason,
  }: {
    class: typeof DomainObject;
    reason: string;
  }) {
    super(
      `
Can not reference declared resource of class '${ofClass.name}'. Instances of the class '${ofClass.name}' can not be referenced because ${reason}.
    `.trim(),
    );
  }
}

export class CanNotBuildReferenceError extends Error {
  constructor({
    referenceInput,
    referenceOf,
    reason,
  }: {
    referenceInput: any;
    referenceOf: any;
    reason: string;
  }) {
    super(
      `
Can not build reference ${
        referenceOf?.name ? `to class '${referenceOf.name}'` : ''
      } because ${reason}.

referenceInput
${JSON.stringify(referenceInput, null, 2)}
    `.trim(),
    );
  }
}

/**
 * build a reference to a resource of a particular class using a supported key
 * - uses the unique key if possible
 * - uses the primary key otherwise
 *
 * note
 * - we prefer unique key over primary key because
 *   - it encodes the most information
 *   - it is always available for resources
 *   - it is the common comparable form of reference
 */
export const buildReferenceTo = <
  /**
   * the type of the domain object
   */
  T extends DomainObject<any>,
  /**
   * the class of the domain object
   */
  C extends new (props: T) => T,
  /**
   * the names of the primary key attributes
   */
  P extends keyof T,
  /**
   * the names of the unique key attributes
   */
  U extends keyof T,
  /**
   * the key we are attempting to build a reference with
   */
  K extends Partial<T>,
>(
  /**
   * the class this is a reference to
   */
  to: C,

  /**
   * the data to build the reference with
   */
  using: K,

  /**
   * the class to instantiate the reference with
   * - defaults to the generic DeclaredResourceReference class
   * - allows using a more specific subclass if one was created
   */
  as?: new (
    props: DeclaredResourceReference<T, P, U>,
  ) => DeclaredResourceReference<T, P, U>,
): DeclaredResourceReference<T, P, U> => {
  const referenceOf = to;
  const referenceInput = using;
  const ReferenceConstructor =
    as ??
    defineReferenceClassOf({ class: referenceOf }) ??
    DeclaredResourceReference; // todo: start throwing errors if reference is not explicitly declared on the source class; using DeclaredResourceError produces hydration and serialization errors for "alternatives" types

  // get the unique and unique key defs for the class
  const { unique, primary } = defineReferenceKeyConstituentsOf({
    class: referenceOf,
  });

  // if unique key is defined, reference by unique key
  const hasUniqueKey = unique.every((key) => key in referenceInput);
  if (hasUniqueKey)
    return new ReferenceConstructor({
      referenceOf: referenceOf.name,
      identifiedBy: {
        key: DeclaredResourceReferenceKeyType.UNIQUE_KEY,
        value: buildUniqueKeyTo(referenceOf, referenceInput),
      },
    }) as DeclaredResourceReference<T, P, U>;

  // if primary key is defined, reference by primary key
  const hasPrimaryKey = primary.every((key) => key in referenceInput);
  if (hasPrimaryKey)
    return new ReferenceConstructor({
      referenceOf: referenceOf.name,
      identifiedBy: {
        key: DeclaredResourceReferenceKeyType.PRIMARY_KEY,
        value: buildPrimaryKeyTo(referenceOf, referenceInput),
      },
    }) as DeclaredResourceReference<T, P, U>;

  // if has neither, then the reference can not be made
  throw new CanNotBuildReferenceError({
    referenceOf,
    referenceInput,
    reason:
      'neither the primary key nor unique key is present in the reference key',
  });
};

export { buildReferenceTo as buildRef };
