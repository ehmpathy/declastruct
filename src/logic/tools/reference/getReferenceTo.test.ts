import { DomainObject } from 'domain-objects';

import { CanNotReferenceDeclaredResourceClassError } from './buildReferenceTo';
import {
  CanNotReferenceDeclaredResourceError,
  getReferenceTo,
} from './getReferenceTo';

// todo: remove the declastruct references in favor of domain-object references
describe.skip('getReferenceTo', () => {
  it('should throw a helpful error if the resource is not an instance of a domain object', () => {
    try {
      getReferenceTo({ name: 'bob', age: 21 });
      throw new Error('should not reach here');
    } catch (error: any) {
      expect(error).toBeInstanceOf(CanNotReferenceDeclaredResourceError);
      expect(error.message).toContain('resource is not an instance of a');
      expect(error).toMatchSnapshot();
    }
  });
  it('should throw a helpful error the class does not have unique key defined', () => {
    interface CNCMachine {
      uuid?: string;
      serialNumber: string;
      size: string;
      location: string;
    }
    class CNCMachine extends DomainObject<CNCMachine> implements CNCMachine {
      public static primary = ['uuid'];
    }
    try {
      getReferenceTo(
        new CNCMachine({
          serialNumber: '821',
          size: 'big',
          location: 'basement',
        }),
      );
      throw new Error('should not reach here');
    } catch (error: any) {
      expect(error).toBeInstanceOf(CanNotReferenceDeclaredResourceClassError);
      expect(error.message).toContain(
        `the static property 'unique' was not defined as an array of strings`,
      );
      expect(error).toMatchSnapshot();
    }
  });
  it('should throw a helpful error the class does not have primary key defined correctly', () => {
    interface CNCMachine {
      uuid?: string;
      serialNumber: string;
      size: string;
      location: string;
    }
    class CNCMachine extends DomainObject<CNCMachine> implements CNCMachine {
      public static primary = 'uuid';
    }
    try {
      getReferenceTo(
        new CNCMachine({
          serialNumber: '821',
          size: 'big',
          location: 'basement',
        }),
      );
      throw new Error('should not reach here');
    } catch (error: any) {
      expect(error).toBeInstanceOf(CanNotReferenceDeclaredResourceError);
      expect(error.message).toContain(
        `the static property 'primary' was not defined as an array of strings`,
      );
      expect(error).toMatchSnapshot();
    }
  });
  it('should be able to reference by unique key', () => {
    interface CNCMachine {
      uuid?: string;
      serialNumber: string;
      size: string;
      location: string;
    }
    class CNCMachine extends DomainObject<CNCMachine> implements CNCMachine {
      public static primary = ['uuid'];
      public static unique = ['serialNumber'];
    }
    const reference = getReferenceTo(
      new CNCMachine({
        serialNumber: '821',
        size: 'big',
        location: 'basement',
      }),
    );
    expect(reference).toEqual({
      referenceOf: 'CNCMachine',
      identifiedBy: {
        key: 'UNIQUE_KEY',
        value: { serialNumber: '821' },
      },
    });
    expect(reference).toMatchSnapshot(); // log for viewing
  });
  it('should reference by primary key if posible', () => {
    interface CNCMachine {
      uuid?: string;
      serialNumber: string;
      size: string;
      location: string;
    }
    class CNCMachine extends DomainObject<CNCMachine> implements CNCMachine {
      public static primary = ['uuid'];
      public static unique = ['serialNumber'];
    }
    const reference = getReferenceTo(
      new CNCMachine({
        uuid: 'de0e34ca-ce61-4884-ace5-093a1a32ff92',
        serialNumber: '821',
        size: 'big',
        location: 'basement',
      }),
    );
    expect(reference).toEqual({
      referenceOf: 'CNCMachine',
      identifiedBy: {
        key: 'PRIMARY_KEY',
        value: { uuid: 'de0e34ca-ce61-4884-ace5-093a1a32ff92' },
      },
    });
    expect(reference).toMatchSnapshot(); // log for viewing
  });
});
