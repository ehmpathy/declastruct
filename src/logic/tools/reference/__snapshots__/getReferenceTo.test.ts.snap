// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getReferenceTo should be able to reference by unique key 1`] = `
Object {
  "identifiedBy": Object {
    "key": "UNIQUE_KEY",
    "value": Object {
      "serialNumber": "821",
    },
  },
  "referenceOf": "CNCMachine",
}
`;

exports[`getReferenceTo should reference by primary key if posible 1`] = `
Object {
  "identifiedBy": Object {
    "key": "PRIMARY_KEY",
    "value": Object {
      "uuid": "de0e34ca-ce61-4884-ace5-093a1a32ff92",
    },
  },
  "referenceOf": "CNCMachine",
}
`;

exports[`getReferenceTo should throw a helpful error if the resource is not an instance of a domain object 1`] = `
[Error: Can not reference declared resource of class 'Object' because the resource is not an instance of a DomainObject class.

resource
{
  "name": "bob",
  "age": 21
}]
`;

exports[`getReferenceTo should throw a helpful error the class does not have primary key defined correctly 1`] = `[Error: Can not reference declared resource of class 'CNCMachine'. Instances of the class 'CNCMachine' can not be referenced because the static property 'primary' was not defined as an array of strings.]`;

exports[`getReferenceTo should throw a helpful error the class does not have unique key defined 1`] = `[Error: Can not reference declared resource of class 'CNCMachine'. Instances of the class 'CNCMachine' can not be referenced because the static property 'unique' was not defined as an array of strings.]`;
